Closes #23

fma1 opened the pull request:
- Added usingAListFieldWithListOfIsImmutable() test for Java 9 factory methods along with CopyListIntoNewListUsingListOf class
- Added property for activation for jdk11 profile

fma1 commented:
@Grundlefleck 

So I ran `Jdk8OptionalUsage` and it seems like it works:
![clientofmutabilitydetector](https://user-images.githubusercontent.com/15847948/87064278-7bf7ec00-c1dd-11ea-8ff1-580175e317bf.png)

Originally since there are different profiles I thought I could just use the jdk11 profile but it seems like the Travis build is specifically checking for JDK 8. I thought changing the jdk8 profile to jdk9 would work but hmm it seems like that doesn't work because the Java 8 compiler is specifically being used.

Grundlefleck commented:
If you're removing jdk8 as a profile I think you'd need to change travis.yml in sync with it.

fma1 commented:
Ah I see. I think it should be good to merge now.

Grundlefleck commented:
Nice!

One other nit-pick I should have said earlier: the test method is in a class whose name doesn't reflect what's being tested. Java9ImmutableCollectionsTest or something like that may be better?

After that I've no nitpicks :)

fma1 commented:
Yeah, I forgot to ask. I meant to ask what you thought about the naming. As Martin Fowler once stated, "There are only two hard things in Computer Science: cache invalidation and naming things." :)

Grundlefleck commented:
The test method name was fine, it was the class name Jdk8OptionalUsage that I meant should change.

fma1 commented:
I rebased and dropped the last commit.

I am actually thinking now because the `Optional` stuff is like half the test class. I am wondering if all of those things should be in the same class? The main reason I put the new test there is because I just did a `grep -r unmodifiable .` when looking for where to put the test.

fma1 commented:
I ended up creating a new class called `JdkImmutableCollectionsTest`. Let me know if you want it changed. Thinking about it now, if I think about `Option` in Scala, I see it as a 1-element collection. So maybe they could all be in one class if you consider `Optional` a collection. I was just thinking of them as separate so the class name would be pretty long.

fma1 commented:
So making sure I was using the latest master for `MutabilityDetector`, I ran `mvn install` and using version `0.10.4-SNAPSHOT`, the test runs successfully. But my guess is Travis is using an older version of the library. Hmm, maybe this might have to wait for release?

Grundlefleck commented:
You're one step ahead of me, I just discovered the same thing.

I'm not sure how Travis can pass on some [runs](https://travis-ci.org/github/MutabilityDetector/ClientOfMutabilityDetector/jobs/706625224), using 0.10.2 but not on others. But like you, I'm convinced the change is good based on running locally, and I'll try to figure out what Travis is doing. I may even replace with GitHub Actions in this repo too.

Grundlefleck commented:
Thanks again @fma1! 

I'll look into the Travis vs. GitHub Actions issue separately.

I'm very pleased with your contributions. How did you find it? Would you like to do more?

fma1 commented:
You're welcome @Grundlefleck :)

> How did you find it?  

I found it a lot of fun but also somewhat challenging at first. But it takes time to get acclimated to a codebase. Actually thinking about it now I should have used the Flow plugin. I used that when I first started working at my company because like the codebase was massive and the code I needed to add functionality to was really old legacy code so I had to reverse engineer how the whole library/module worked.

A little background about me is I started programming when I was 8 years old. So to me programming is sort of like legos. With legos you put them together one by one and then eventually you get something like a castle. And when I was kid I just played with legos for fun. Programming was the same to me. I think some people in the industry are employed simply for the monetary value, and I respect that, but to me, I truly enjoy it.

I also possess a CS degree and the class I enjoyed the most was Compilers. I just think [Compiler Theory](https://www.youtube.com/watch?v=Qkwj65l_96I&list=PLEbnTDJUr_IcPtUXFy2b1sGRPsLFMghhS) is really interesting stuff. Thinking about it, in order to detect a reflective call is equivalent to a non-reflective call, you would have to analyze things at runtime (dynamic analysis). Static analysis wouldn't work in that case.

> Would you like to do more?

Sure :)
